scala/data_structures/bloom_filter.scala

```diff
@@ -1,66 +1,78 @@
 package data_structures
 
+import scala.util.hashing.MurmurHash3.stringHash
+import Math.{abs, ceil, log, pow, round}
+
 case class BloomFilter[A](
   nbOfItems: Int,
   falsePositiveProbability: Float,
   private val array: Array[Boolean],
-  private val hashFunctions: Array[String => Int]
+  private val numberOfHashFunctions: Int,
+  private val hashSeed: Int
 ) {
 
+  private val hashFunctions: Stream[String => Int] =
+    (1 to numberOfHashFunctions)
+      .toStream
+      .map(i => (str: String) => abs(stringHash(str, hashSeed + i)) % array.length)
+
   def +=(item: A): BloomFilter[A] = {
     val itemString = item.toString
-    val hashes = this.hashFunctions.map(_(itemString))
-    val updatedArray = hashes.foldLeft(this.array)((acc, hash) => acc.updated(hash, true))
-    BloomFilter(nbOfItems, falsePositiveProbability, updatedArray, hashFunctions)
+    val hashes = hashFunctions.map(_(itemString))
+    val updatedArray = hashes.foldLeft(array)((acc, hash) => acc.updated(hash, true))
+    this.copy(array=updatedArray)
```

Comment:
Learned about this trick to update only 1 property of a case class

Reply:
So this means that the entire array is copied on each insertion? Is this optimized away somehow by Scala? it seems like it will affect performance bigly

Reply:
No the array isn't being copieid. Because Scala data structures are immutable, that only way for me to modify my BloomFilter instance is to create a copy of it.
So this method creates a copy of the current `BloomFilter` instance, with only one difference: a new array.
However, `updatedArray` is indeed a copy of `array`, because that's how fold works and that's the essence of immutable data structures: you can only create new ones, not modify the existing.
And your hunch was right, copying immutable data structures is heavily optimized to be as close as possible to mutating data structures as possible. You can watch this if you're interested in learning more: https://www.youtube.com/watch?v=Wo0qiGPSV-s